### ISSUE TITLE
NineSlice taints predefined layout definitions when used by addons

### ISSUE THREAD

**Issue by Meorawr** [Stanzilla/WoWUIBugs#107]
Addons that make use of specific FrameXML templates that apply predefined nineslice layouts can spread taint to further uses of those layouts when used by Blizzard addons loaded later in the session. This issue was partially discovered during the 9.0 beta (#28) however the fix applied for that didn't fundamentally fix the issue described herein.

---

This is most easily reproduced through any template that derives from ButtonFrameTemplate.

1. Enable the maximum level of taint logging via `/console taintLog 11`.
2. Run the following snippet to create a frame with a taintable layout: `/run local Frame = CreateFrame("Frame", nil, UIParent, "ButtonFrameTemplate"); ButtonFrameTemplate_HidePortrait(Frame);`
3. Open the event log via `/etrace`.
4. Disable taint logging via `/console taintLog 0`.

The relevant part of the generated taint log will point to the "mirrorLayout" field which is both read and modified by the [PropagateLayoutSettingsToPieceLayout](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/SharedXML/NineSlice.lua#L81) function.

```
4/22 00:49:04.091  Global variable mirrorLayout tainted by MACRO_TAINT - Interface\SharedXML\NineSlice.lua:84 PropagateLayoutSettingsToPieceLayout()
4/22 00:49:04.091      Interface\SharedXML\NineSlice.lua:400 ApplyLayout()
4/22 00:49:04.091      Interface\SharedXML\PortraitFrame.lua:29 <unnamed>:SetBorder()
4/22 00:49:04.091      Interface\SharedXML\SharedUIPanelTemplates.lua:102 ButtonFrameTemplate_HidePortrait()
4/22 00:49:04.091      local Frame = CreateFrame("Frame", nil, UIParent, "ButtonFrameTemplate"); ButtonFrameTemplate_HidePortrait(Frame);:1
4/22 00:49:04.091      RunScript()
4/22 00:49:04.091      Interface\FrameXML\ChatFrame.lua:2194 ?()
4/22 00:49:04.091      Interface\FrameXML\ChatFrame.lua:5201 ChatEdit_ParseText()
4/22 00:49:04.091      Interface\FrameXML\ChatFrame.lua:4865 ChatEdit_SendText()
4/22 00:49:04.091      Interface\FrameXML\ChatFrame.lua:4901 ChatEdit_OnEnterPressed()
4/22 00:49:04.091      ChatFrame1EditBox:OnEnterPressed()

...

4/22 00:49:05.372  Execution tainted by MACRO_TAINT while reading mirrorLayout - Interface\SharedXML\NineSlice.lua:83 PropagateLayoutSettingsToPieceLayout()
4/22 00:49:05.372      Interface\SharedXML\NineSlice.lua:400 ApplyLayout()
4/22 00:49:05.372      Interface\SharedXML\PortraitFrame.lua:29 EventTrace:SetBorder()
4/22 00:49:05.372      Interface\SharedXML\SharedUIPanelTemplates.lua:102 ButtonFrameTemplate_HidePortrait()
4/22 00:49:05.372      Interface\AddOns\Blizzard_EventTrace\Blizzard_EventTrace.lua:75
4/22 00:49:05.372      LoadAddOn()
4/22 00:49:05.372      Interface\FrameXML\UIParent.lua:509 UIParentLoadAddOn()
4/22 00:49:05.372      Interface\FrameXML\ChatFrame.lua:2401 ?()
4/22 00:49:05.372      Interface\FrameXML\ChatFrame.lua:5201 ChatEdit_ParseText()
4/22 00:49:05.372      Interface\FrameXML\ChatFrame.lua:4865 ChatEdit_SendText()
4/22 00:49:05.372      Interface\FrameXML\ChatFrame.lua:4901 ChatEdit_OnEnterPressed()
4/22 00:49:05.372      ChatFrame1EditBox:OnEnterPressed()
```

The sequence of events that occurs is as follows:

**Insecure code**

1. We create a frame that derives from ButtonFrameTemplate.
2. We call [ButtonFrameTemplate_HidePortrait](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/SharedXML/SharedUIPanelTemplates.lua#L101) which sets the frames' nineslice layout to "ButtonFrameTemplateNoPortrait".
3. When applying the layout we reach the [PropagateLayoutSettingsToPieceLayout](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/SharedXML/NineSlice.lua#L81) function which reads the `mirrorLayout` field on the piece layout table. This will be nil, as the specified layout doesn't have a per-piece override for that attribute.
4. As the field is nil the code now attempts to assign it any existing (and possibly nil) value from the overall table for the layout. This assignment taints, which will cause all future reads from its enclosing branch to always taint execution for this layout.

**Secure code**

5. The event trace tool is loaded, for which its frame inherits ButtonFrameTemplate/PortraitFrameTemplate.
6. The event trace tool calls [ButtonFrameTemplate_HidePortrait](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/SharedXML/SharedUIPanelTemplates.lua#L101) and sets its nineslice layout to "ButtonFrameTemplateNoPortrait".
7. When applying the layout we reach the [PropagateLayoutSettingsToPieceLayout](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/SharedXML/NineSlice.lua#L81) function and read the now-tainted `mirrorLayout` field on each piece, tainting all further execution.

---

There's two simple solutions depending on how you particularly feel about the matter.

**Option 1** 

This is the smaller of the fixes and should work well enough; if the call path isn't secure then the piece layout table should be copied prior to propagating layout settings.

In a pristine secure environment this has no impact elsewhere, and only when addons are involved does the table copy potentially occur - and only on paths that are tainted to begin with. The performance cost of this in the insecure code case should be minimal as I don't believe nineslice layout applications are all that common outside of initial frame layout.

```diff
diff --git a/SharedXML/NineSlice.lua b/SharedXML/NineSlice.lua
index b560fd53..efc9de4f 100644
--- a/SharedXML/NineSlice.lua
+++ b/SharedXML/NineSlice.lua
@@ -397,6 +397,11 @@ function NineSliceUtil.ApplyLayout(container, userLayout, textureKit)
 		local pieceName = setup.pieceName;
 		local pieceLayout = userLayout[pieceName];
 		if pieceLayout then
+
+			if not issecure() then
+				pieceLayout = CopyTable(pieceLayout, true);
+			end
+
 			PropagateLayoutSettingsToPieceLayout(userLayout, pieceLayout);
 
 			local piece, pieceAlreadyExisted = GetNineSlicePiece(container, pieceName);
```

**Option 2**

This option is more thorough and completely eliminates the taint by just removing the whole concept of field propagation and instead modifying the one location that the `mirrorLayout` field is checked to instead manually consult both the piece-specific and overall-layout tables.

This option may have an impact on external code that's making use of the custom piece visual setup function; the only use of it within the public FrameXML code is for backdrops which aren't affected. Addons _could_ be impacted but I'd wager that it'd be an extremely small risk.

1. Add the `userLayout` table as an extra parameter supplied to both the [userLayout.setupPieceVisualsFunction](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/SharedXML/NineSlice.lua#L411) and [SetupPieceVisuals](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/SharedXML/NineSlice.lua#L413) calls. 
2. Update the [SetupPieceVisuals](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/SharedXML/NineSlice.lua#L110) function and pass the new `userLayout` parameter through to [SetupTextureCoordinates](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/SharedXML/NineSlice.lua#L90).
3. Modify [this single usage of the `mirrorLayout` field](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/SharedXML/NineSlice.lua#L93) to check both the `pieceLayout` and `userLayout` tables.

```diff
diff --git a/SharedXML/NineSlice.lua b/SharedXML/NineSlice.lua
index b560fd53..a231a9b1 100644
--- a/SharedXML/NineSlice.lua
+++ b/SharedXML/NineSlice.lua
@@ -78,19 +78,10 @@ local function GetNineSlicePiece(container, pieceName)
 	end
 end
 
-local function PropagateLayoutSettingsToPieceLayout(userLayout, pieceLayout)
-	-- Only apply mirrorLayout if it wasn't explicitly defined
-	if pieceLayout.mirrorLayout == nil then
-		pieceLayout.mirrorLayout = userLayout.mirrorLayout;
-	end
-
-	-- ... and other settings that apply to the whole nine-slice
-end
-
-local function SetupTextureCoordinates(piece, setupInfo, pieceLayout)
+local function SetupTextureCoordinates(piece, setupInfo, pieceLayout, userLayout)
 	local left, right, top, bottom = 0, 1, 0, 1;
 
-	if pieceLayout.mirrorLayout then
+	if pieceLayout.mirrorLayout or (pieceLayout.mirrorLayout ~= false and userLayout.mirrorLayout) then
 		if setupInfo.mirrorVertical then
 			top, bottom = bottom, top;
 		end
@@ -105,9 +96,9 @@ local function SetupTextureCoordinates(piece, setupInfo, pieceLayout)
 	piece:SetTexCoord(left, right, top, bottom);
 end
 
-local function SetupPieceVisuals(piece, setupInfo, pieceLayout, textureKit)
+local function SetupPieceVisuals(piece, setupInfo, pieceLayout, textureKit, userLayout)
 	--- Change texture coordinates before applying atlas.
-	SetupTextureCoordinates(piece, setupInfo, pieceLayout);
+	SetupTextureCoordinates(piece, setupInfo, pieceLayout, userLayout);
 
 	-- textureKit is optional, that's fine; but if it's nil the caller should ensure that there are no format specifiers in .atlas
 	local atlasName = GetFinalNameFromTextureKit(pieceLayout.atlas, textureKit);
@@ -397,8 +388,6 @@ function NineSliceUtil.ApplyLayout(container, userLayout, textureKit)
 		local pieceName = setup.pieceName;
 		local pieceLayout = userLayout[pieceName];
 		if pieceLayout then
-			PropagateLayoutSettingsToPieceLayout(userLayout, pieceLayout);
-
 			local piece, pieceAlreadyExisted = GetNineSlicePiece(container, pieceName);
 			if not pieceAlreadyExisted then
 				container[pieceName] = piece;
@@ -408,9 +397,9 @@ function NineSliceUtil.ApplyLayout(container, userLayout, textureKit)
 			-- Piece setup can change arbitrary properties, do it before changing the texture.
 			setup.fn(container, piece, setup, pieceLayout);
 			if userLayout.setupPieceVisualsFunction then
-				userLayout.setupPieceVisualsFunction(container, piece, setup, pieceLayout, textureKit);
+				userLayout.setupPieceVisualsFunction(container, piece, setup, pieceLayout, textureKit, userLayout);
 			else
-				SetupPieceVisuals(piece, setup, pieceLayout, textureKit);
+				SetupPieceVisuals(piece, setup, pieceLayout, textureKit, userLayout);
 			end
 		end
 	end
```

**Comment by Meorawr** [comment]
Should be fixed in an upcoming 2.5.1 and 9.1 build.

**Comment by Meorawr** [comment]
For TBC this is resolved in 2.5.1.38453.

**Comment by Meorawr** [comment]
For Retail this is resolved in 9.1.0.38511.
